internal/api/pkce.go

```diff
@@ -0,0 +1,25 @@
+package api
+
+import (
+	"regexp"
+)
+
+const PKCE = "pkce"
+
+const MinCodeChallengeLength = 43
+const MaxCodeChallengeLength = 128
+
+var codeChallengePattern = regexp.MustCompile("^[a-zA-Z-._~0-9]+$")
+
+func isValidCodeChallenge(codeChallenge string) (bool, error) {
+	// See RFC 7636 Section 4.2: https://www.rfc-editor.org/rfc/rfc7636#section-4.2
+	hasValidChallengeChars := codeChallengePattern.MatchString
+	switch codeChallengeLength := len(codeChallenge); {
+	case codeChallengeLength < MinCodeChallengeLength, codeChallengeLength > MaxCodeChallengeLength:
+		return false, badRequestError("code challenge has to be between %v and %v characters", MinCodeChallengeLength, MaxCodeChallengeLength)
+	case !hasValidChallengeChars(codeChallenge):
+		return false, badRequestError("code challenge can only contain alphanumeric characters, hyphens, periods, underscores and tildes")
```

Comment:
```suggestion
	// See RFC 7636 Section 4.2: https://www.rfc-editor.org/rfc/rfc7636#section-4.2
	
	switch codeChallengeLength := len(codeChallenge); {
	case codeChallengeLength < MinCodeChallengeLength, codeChallengeLength > MaxCodeChallengeLength:
		return false, badRequestError("code challenge has to be between %v and %v characters", MinCodeChallengeLength, MaxCodeChallengeLength)
	case !codeChallengePattern.MatchString(codeChallenge):
		return false, badRequestError("code challenge can only contain alphanumeric characters, hyphens, periods, underscores and tildes")
```

Why not this? No need to put the function into a variable.